{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File Explanation\n",
    "\n",
    "**loadData.ipynb:**\n",
    "<br> This notebook is to load raw AddNeuroMed data from 'mendelianGenesX2_v3.Rdata' and turn into NumPy arrays and lists for computation advantage \n",
    "\n",
    "**Processes are as follows:**\n",
    "<br> 1) Load all variables from 'mendelianX2_v3.Rdata'\n",
    "<br> 2) Store most of them as NumPy arrays by using \"rpy2\" which converts R data into either NumPy or Pandas \n",
    "<br> 3) Store other variables (nnNs_lRlN, nPairs_lPlC) as lists\n",
    "<br> 4) Load \".bed, .bim, .fam\" files of AddNeuroMed\n",
    "<br> 5) Retrieve SNP locations from .bim\n",
    "<br> 6) Get gene start and end points by using **Biopython package's Entrez API**\n",
    "<br> 7) Store necessary files in .csv format for MatrixEQTL and LD analysis\n",
    "<br> 8) Save them in \"loadData.pickle\" format to be called by \"preprocessData.ipynb\"\n",
    "\n",
    "**Some abbreviations used:**\n",
    "<br> **S:** number of subjects \n",
    "<br> **D:** number of demographic category\n",
    "<br> **G:** number of genes\n",
    "<br> **R:** number of transcripts\n",
    "<br> **N:** number of SNPs\n",
    "<br> **P:** number of pairs\n",
    "<br> **C:** number of categories \n",
    "\n",
    "**Variables created:**\n",
    "<br> 1) **sDem_nSD:** Demographic information of subjects, a NumPy array, shape of (S x D) \n",
    "<br> 2) **sGenesAD_nG:** Genes, a NumPy array, shape of (G x 1)\n",
    "<br> 3) **rRna_nSR:** RNA (gene) expression levels, a NumPy array, shape of (S x R)\n",
    "<br> 4) **rSnp_nSN:** SNP encodings (1, 2, 3), a NumPy array, shape of (S x N)\n",
    "<br> 5) **nNs_lRlN:** SNP names for corresponding genes on which SNPs are, a list, length of (R : [N]) \n",
    "<br> 6) **nPairs_lPlC:** Known gene pairs, a list, length of (2204) with columns for \"TF ID, TF Entrez ID, TF Gene Symbol, Target Entrez ID, Target Gene Symbol\"\n",
    "<br> 7) **nSs:** Subject names\n",
    "<br> 8) **nRs:** Transcript (gene) names\n",
    "<br> 9) **nDs:** Demographic category names\n",
    "<br> 10) **nNs:** SNPs names \n",
    "\n",
    "**Files created for MatrixEQTL + LD analysis:**\n",
    "<br> 1) **eqtl_expression.csv**: Gene expressions of all transcripts, csv format (rows: R, columns: S)\n",
    "<br> 2) **eqtl_genotype.csv**: SNP encodings (0, 1, 2) of SNPs, csv format (rows: N, columns: S)\n",
    "<br> 3) **eqtl_probe_loc.csv**: Gene locations, csv format (rows: R, columns: gene ID, chromosome number, start location, end location)\n",
    "<br> 4) **eqtl_snp_loc.csv**: SNP locations, csv format (rows: N, columns: SNP ID, chromosome number, position)\n",
    "<br> 5) **eqtl_subjects.csv**: Subject names, csv format (rows: S, columns: 1)\n",
    "<br> 6) **eqtl_snps.csv**: SNP names, csv format (rows: N, columns: 1)\n",
    "<br> 7) **eqtl_genes.csv**: Transcript names, csv format (rows: R, columns: 1)\n",
    "<br> 8) **eqtl_snp_gene.csv**: SNP to gene mapping, csv format (rows: N, columns: SNP ID, gene ID)\n",
    "<br> 9) **eqtl_snp_alleles.csv**: SNP allele pairs, csv format (rows: N, columns: a0, a1 for minor and major allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import numpy as np\n",
    "import tensorflow as tf\n",
    "from tensorflow.contrib import rnn\n",
    "import os\n",
    "\n",
    "# Set up environment variables for rpy2 to work\n",
    "os.environ['R_HOME'] = '/usr/local/Cellar/r/3.6.0_2/lib/R'\n",
    "os.environ['R_USER'] = '/usr/local/lib/python3.7/site-packages/rpy2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data loaded from Rdata file.\n",
      "Variables loaded:\n",
      "dem_iSiD\n",
      "demRnaSnp_rc_iSiG\n",
      "genesAD_iG\n",
      "nnInput_jPiNiF\n",
      "nNs_c5_jG\n",
      "pairs_c5_PiC\n",
      "rna_iSiR\n",
      "rna_iSiR_p\n",
      "snp_iSiN\n"
     ]
    }
   ],
   "source": [
    "# Load Rdata with rpy2\n",
    "import rpy2\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "pandas2ri.activate()\n",
    "# Clear the workspace\n",
    "r( 'rm( list = ls( ) )' )\n",
    "# 16GB memory limit\n",
    "r( 'memory.limit( 16000 )' )\n",
    "r['load']('mendelianGenesX2_v3.Rdata')\n",
    "print( 'Data loaded from Rdata file.' )\n",
    "\n",
    "# Hale the user\n",
    "print( 'Variables loaded:' )\n",
    "[ print( a ) for a in r['ls']() ];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Dimensions\n",
    "assert( all( r( 'dimnames( snp_iSiN )[[1]] == dimnames( rna_iSiR )[[1]]' ) ) )\n",
    "assert( all( r( 'names( nNs_c5_jG ) == gsub( \"rna\", \"\", dimnames( rna_iSiR )[[2]] )' )  ) )\n",
    "nSs = r( 'dimnames( rna_iSiR )[[1]]' ) \n",
    "nRs = r( 'dimnames( rna_iSiR )[[2]]' ) \n",
    "nDs = r( 'dimnames( dem_iSiD )[[2]]' ) \n",
    "nNs = r( 'dimnames( snp_iSiN )[[2]]' ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TRANSFORM DATA INTO NUMPYS\n",
    "# Transforms all the data from rpy2 objects into NumPys and lists\n",
    "\n",
    "# Transform R objects into numpys\n",
    "import numpy as np\n",
    "## dem_iSiD\n",
    "sDem_nSD = np.array( rpy2.robjects.r( 't( dem_iSiD )' ),\n",
    "                    dtype = np.str )\n",
    "sDem_nSD.shape = r( 'dim( dem_iSiD )' )\n",
    "assert( all( sDem_nSD[0:10,10] == ( r( 'dem_iSiD[1:10,11]' ) ) ) )\n",
    "\n",
    "## genesAD_iG\n",
    "sGenesAD_nG = np.array( rpy2.robjects.r( 'genesAD_iG' ),\n",
    "                    dtype = np.str )\n",
    "sGenesAD_nG = np.transpose( sGenesAD_nG )\n",
    "assert( all( sGenesAD_nG[0:10] == ( r( 'genesAD_iG[1:10]' ) ) )  ) \n",
    "\n",
    "## rna_iSiR\n",
    "rRna_nSR = np.array( rpy2.robjects.r( 'rna_iSiR' ),\n",
    "                     dtype = np.float )\n",
    "assert( all( rRna_nSR[0:10,10] == ( r( 'rna_iSiR[1:10,11]' ) ) ) )\n",
    "\n",
    "## snp_iSiN\n",
    "rSnp_nSN = np.array( rpy2.robjects.r( 'snp_iSiN' ),\n",
    "                     dtype = np.float )\n",
    "rSnp_nSN.shape = r( 'dim( snp_iSiN )' )\n",
    "assert( all( rSnp_nSN[0:100,10] == [ nSRN for nSRN in map(float, (r( 'snp_iSiN[1:100,11]' ) ) ) ] ) )\n",
    "\n",
    "## nNs_c5_jG \n",
    "nNs_lRlN = [  nNs  for nNs in r( 'nNs_c5_jG') ]\n",
    "assert(all( (nNs_lRlN[0][0:10]) == ( r( 'nNs_c5_jG[[1]][1:10]' ) ) ) )\n",
    "assert(all( (nNs_lRlN[10][0:10]) == ( r( 'nNs_c5_jG[[11]][1:10]' ) ) ) )\n",
    "\n",
    "## pairs_c5_PiC\n",
    "nPairs_lPlC = [  nPairs  for nPairs in r( 'pairs_c5_PiC') ]\n",
    "assert(all( (nPairs_lPlC[0][0:10]) == ( r( 'pairs_c5_PiC[[1]][1:10]') ) ) )\n",
    "assert(all( (nPairs_lPlC[4][0:10]) == ( r( 'pairs_c5_PiC[[5]][1:10]' ) ) ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data transformed into numpys.\n"
     ]
    }
   ],
   "source": [
    "# Check data\n",
    "assert( len( nNs_lRlN ) == rRna_nSR.shape[1] )\n",
    "assert( sDem_nSD.shape[0] == rRna_nSR.shape[0] )\n",
    "assert( sDem_nSD.shape[0] == rSnp_nSN.shape[0] )\n",
    "print( 'Data transformed into numpys.' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:21<00:00,  9.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotypes loaded\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# LOAD HBTRC STUDY DATA\n",
    "from pandas_plink import read_plink1_bin, read_plink\n",
    "import pandas as pd\n",
    "# read_plink1_bin returns (dataArray in shape = (samples, variants))\n",
    "\n",
    "# vBim_pNC: allele, vFam_pSC: samples, vBed_aNS: genotype (snp)\n",
    "(vBim_pNC, vFam_pSC, vBed_aNS) = read_plink(\"ANM-files/anm_batch1_batch2_merged\")\n",
    "print( \"Genotypes loaded\" )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chrom</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a0</th>\n",
       "      <th>a1</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs143840423</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2936385</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs12564456</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2936386</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs182722896</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2936420</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs72629483</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2936678</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs55949537</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2936746</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           snp chrom   cm      pos a0 a1     i\n",
       "0  rs143840423     1  0.0  2936385  T  C  1881\n",
       "1   rs12564456     1  0.0  2936386  G  C  1882\n",
       "2  rs182722896     1  0.0  2936420  C  A  1883\n",
       "3   rs72629483     1  0.0  2936678  C  T  1884\n",
       "4   rs55949537     1  0.0  2936746  T  A  1885"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Columns:\n",
    "\n",
    "Chrom: Chromosome code (either an integer, or 'X'/'Y'/'XY'/'MT'; '0' indicates unknown) or name\n",
    "SNP: Variant identifier\n",
    "Cm: Position in morgans or centimorgans (safe to use dummy value of '0')\n",
    "Pos: Base-pair coordinate (1-based; limited to 2^31-2)\n",
    "A0: Allele 1 (corresponding to clear bits in .bed; usually minor)\n",
    "A1: Allele 2 (corresponding to set bits in .bed; usually major)\n",
    "\"\"\"\n",
    "\n",
    "vBim_pNC = vBim_pNC.set_index('snp')\n",
    "vBim_pNC = vBim_pNC.loc[nNs.tolist()]\n",
    "vBim_pNC.reset_index(inplace=True)\n",
    "vBim_pNC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Find Chromosome locations of AddNeuroMed examples\n",
      "Location search done.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "# FIND IN WHICH GENES GIVEN SNPS ARE\n",
    "nGs = np.array(list(map(lambda s: s.replace('rna' , ''), nRs)))\n",
    "\n",
    "# Find Chromosome locations of AddNeuroMed examples\n",
    "print( \"Find Chromosome locations of AddNeuroMed examples\" )\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"email address\"\n",
    "iChromosome_nX = np.full( shape = nGs.shape,\n",
    "                          fill_value = np.nan )\n",
    "\n",
    "# Request data for all your genes  \n",
    "request = Entrez.epost( \"gene\",\n",
    "                        id = \",\".join( nGs ) )\n",
    "result = Entrez.read(request)\n",
    "webEnv = result[ \"WebEnv\" ]\n",
    "queryKey = result[ \"QueryKey\" ]\n",
    "data = Entrez.esummary( db = \"gene\", \n",
    "                        webenv = webEnv, \n",
    "                       query_key = queryKey )\n",
    "annotations = Entrez.read( data )\n",
    "\n",
    "# Wrangle chromosome locations (uid: unique identifier)\n",
    "sGeneID_0_nG = np.array( [ gene.attributes[ \"uid\" ] for gene in annotations[ \"DocumentSummarySet\" ][ \"DocumentSummary\" ] ] )\n",
    "iChromosome_0_nG = np.array( [ gene[ \"Chromosome\" ] for gene in annotations[ \"DocumentSummarySet\" ][ \"DocumentSummary\" ] ] )\n",
    "assert np.all( nGs == sGeneID_0_nG )\n",
    "\n",
    "# The start and ending locations of each gene on chromosome\n",
    "iChrStart_0_nG = np.array( [ gene[\"GenomicInfo\"][0][\"ChrStart\"] for gene in annotations[ \"DocumentSummarySet\" ][ \"DocumentSummary\" ] ] )\n",
    "iChrStop_0_nG = np.array( [ gene[\"GenomicInfo\"][0][\"ChrStop\"] for gene in annotations[ \"DocumentSummarySet\" ][ \"DocumentSummary\" ] ] )\n",
    "\n",
    "print( \"Location search done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD ELINK SNP_GENE FILE FROM NCBI \n",
    "import numpy as np\n",
    "sSNP_Gene = np.loadtxt(fname=\"/Elink/snp_genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take snps and genes in separate numpy arrays\n",
    "sSNP_nN = sSNP_Gene[:,0]\n",
    "sGene_nN = sSNP_Gene[:,1]\n",
    "\n",
    "sSNP_nN = sSNP_nN.astype('int64')\n",
    "sGene_nN = sGene_nN.astype('int64')\n",
    "\n",
    "# Sort snp names to make search faster\n",
    "perm = np.argsort(sSNP_nN)\n",
    "sSNP_nN = sSNP_nN[perm]\n",
    "sGene_nN = sGene_nN[perm]\n",
    "\n",
    "# SNP IDs given in the vBim_pNC['snp'] column\n",
    "nNn = np.array( [ ( int(vBim_pNC['snp'][i].replace('rs', ''))) \n",
    "                 for i in range(len(vBim_pNC))]) \n",
    "\n",
    "# FIND IN WHICH GENES GIVEN SNPS ARE\n",
    "nGeneID_0_nG = sGene_nN[np.searchsorted(sSNP_nN, nNn)]\n",
    "sGeneID_0_nG = nGeneID_0_nG.astype('str')\n",
    "\n",
    "assert(sGeneID_0_nG.shape[0] == vBim_pNC.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CREATE INPUTS FOR EQTL ANALYSIS\n",
    "genotype = np.zeros(shape = rSnp_nSRN.shape)\n",
    "genotype[rSnp_nSRN == 1] = 0\n",
    "genotype[rSnp_nSRN == 2] = 1\n",
    "genotype[rSnp_nSRN == 3] = 2\n",
    "\n",
    "# GENOTYPE\n",
    "eqtl_genotype = pd.DataFrame(data = genotype.transpose(), columns=nSs)\n",
    "eqtl_genotype.set_index(nNs, inplace=True)\n",
    "eqtl_genotype.index.rename(\"snp\", inplace=True)\n",
    "\n",
    "# GENE EXPRESSION\n",
    "eqtl_expression = pd.DataFrame(data = rRna_nSR.transpose(), columns = nSs)\n",
    "eqtl_expression.set_index(nGs, inplace=True)\n",
    "eqtl_expression.index.rename(\"id\", inplace=True)\n",
    "\n",
    "# PROBE_LOC: GENE LOCATIONS\n",
    "eqtl_probe_loc = pd.DataFrame({'id': nGs,\n",
    "                           'chromosome': iChromosome_0_nG,\n",
    "                           'start': iChrStart_0_nG,\n",
    "                           'end': iChrStop_0_nG\n",
    "                          })\n",
    "\n",
    "# SNP_LOC: SNP POSITIONS\n",
    "eqtl_snp_loc = pd.DataFrame({'snp': vBim_pNC['snp'], \n",
    "                             'chrom_snp': vBim_pNC['chrom'],\n",
    "                             'pos': vBim_pNC['pos'] } )\n",
    "\n",
    "# SNP_ALLELES: SNP ALLELE\n",
    "snp_alleles = pd.DataFrame({'snp': vBim_pNC['snp'], \n",
    "                            'a0': vBim_pNC['a0'],\n",
    "                            'a1': vBim_pNC['a1'] } )\n",
    "\n",
    "# MAPPING from SNP to GENE\n",
    "snp_gene_data = {'snp': nNs, 'gene': sGeneID_0_nG}\n",
    "snp_gene = pd.DataFrame(data = snp_gene_data)\n",
    "\n",
    "assert( eqtl_genotype.shape[1] == eqtl_expression.shape[1] )\n",
    "assert( eqtl_genotype.shape[0] == eqtl_snp_loc.shape[0] )\n",
    "assert( eqtl_expression.shape[0] == eqtl_probe_loc.shape[0] )\n",
    "\n",
    "print(\"eQTL inputs in pandas format created.\")\n",
    "\n",
    "# eQTL: STORE IN .CSV FORMAT\n",
    "expression_csv = eqtl_expression.to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_ANM_expression.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "genotype_csv = eqtl_genotype.to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_ANM_genotype.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "probe_loc_csv = eqtl_probe_loc.to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_ANM_probe_loc.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "snp_loc_csv = eqtl_snp_loc.to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_ANM_snp_loc.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "print(\"eQTL inputs stored in CSV format.\")\n",
    "\n",
    "\n",
    "# HELPER VARIABLES\n",
    "subjects_csv = pd.DataFrame(data=nSs).to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_ANM_subjects.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "snps_csv = pd.DataFrame(data=nNs).to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_ANM_snps.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "genes_csv = pd.DataFrame(data=nGs).to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_ANM_genes.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "snp_gene_csv = pd.DataFrame(data=snp_gene).to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_ANM_snp_gene.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "snp_alleles_csv = pd.DataFrame(data=snp_alleles).to_csv(\"../eQTL + LD analysis/AddNeuroMed/eqtl_snp_alleles.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "print(\"eQTL input helpers stored in CSV format.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data saved into pickle.\n"
     ]
    }
   ],
   "source": [
    "# SAVE DATA\n",
    "# Save all the transformed data into a pickle, such that it can be more easily loaded from Python\n",
    "\n",
    "# Save data into Python file\n",
    "import pickle\n",
    "r( 'rm( list = ls( ) )' ) #Free space\n",
    "with open( 'loadData.pickle', 'wb' ) as f:\n",
    "    pickle.dump( rRna_nSR, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( sDem_nSD, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( rSnp_nSN, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( sGenesAD_nG, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nNs_lRlN, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nPairs_lPlC, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nSs, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nRs, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nDs, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nNs, f, pickle.HIGHEST_PROTOCOL )\n",
    "    print( 'Data saved into pickle.' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
