{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File Explanation\n",
    "\n",
    "**loadData_HBTRC.ipynb:**\n",
    "<br> This notebook is to load raw HBTRC data in '.bim, .bed, and .bam' formats and turn into NumPy arrays and lists for computation advantage \n",
    "\n",
    "**Processes are as follows:**\n",
    "<br>1) Load data from \".bed, .bim, and .fam\" formats by using **\"pandas_plink\"** package\n",
    "<br>2) Genes that are associated with SNPs (by using sSNP_Gene variable) are found and stored as **sGeneID_0_nG**.\n",
    "<br>3) All transcription factors/targets (including Entrez ID and gene ID) in the dataset are identified by \"gwasPreprocessing_v2_tfs.Rdata\".\n",
    "<br>4) Pairs of transcription factors and targets are stored in nPairs_lPlC (function: **tf_Pairs_Finding** in **loadData_utils.py**)\n",
    "<br>5) Missing values in RNA expression data are replaced with average values of particular gene.\n",
    "<br>6) Missing values in SNPs are replaced with 3s as encoding 3 represents the most common allele, \"A/A\".\n",
    "<br>7) Genotypes, gene expression, probe location, SNP location .csv files are created for MatrixEQTL + LD analysis\n",
    "<br> 8) Save them in \"loadData_HBTRC.pickle\" format to be called by \"preprocessData_HBTRC.ipynb\"\n",
    "\n",
    "**Some abbreviations used:**\n",
    "<br> **S:** number of subjects \n",
    "<br> **R:** number of transcripts\n",
    "<br> **N:** number of SNPs\n",
    "<br> **P:** number of pairs\n",
    "<br> **C:** number of categories \n",
    "\n",
    "**Variables created for preprocessData:**\n",
    "<br> 1) **rRna_nSR:** RNA (gene) expression levels, a NumPy array, shape of (S x R)\n",
    "<br> 2) **rSnp_nSN:** SNP encodings, a NumPy array, shape of (S x N)\n",
    "<br> 3) **nNs_lRlN:** SNP names for corresponding genes on which SNPs are, a list, length of (R) \n",
    "<br> 4) **nPairs_lPlC:** Known gene pairs, a list, length of (2204) with columns for \"TF ID, TF Entrez ID, TF Gene Symbol, Target Entrez ID, Target Gene Symbol\"\n",
    "<br> 5) **nSs:** Subject names\n",
    "<br> 6) **nRs:** Transcript (gene) names\n",
    "<br> 7) **nNs:** SNPs names\n",
    "\n",
    "**Files created for MatrixEQTL + LD analysis:**\n",
    "<br> 1) **eqtl_expression.csv**: Gene expressions of all transcripts, csv format (rows: R, columns: S)\n",
    "<br> 2) **eqtl_genotype.csv**: SNP encodings (1, 2, 3) of SNPs, csv format (rows: N, columns: S)\n",
    "<br> 3) **eqtl_probe_loc.csv**: Gene locations, csv format (rows: R, columns: gene ID, chromosome number, start location, end location)\n",
    "<br> 4) **eqtl_snp_loc.csv**: SNP locations, csv format (rows: N, columns: SNP ID, chromosome number, position)\n",
    "<br> 5) **linkdis_genotype.csv**: SNP encodings (0, 1, 2) of SNPs, csv format (rows: N, columns: S)\n",
    "<br> 6) **eqtl_subjects.csv**: Subject names, csv format (rows: S, columns: 1)\n",
    "<br> 7) **eqtl_snps.csv**: SNP names, csv format (rows: N, columns: 1)\n",
    "<br> 8) **eqtl_genes.csv**: Transcript names, csv format (rows: R, columns: 1)\n",
    "<br> 9) **eqtl_snp_gene.csv**: SNP to gene mapping, csv format (rows: N, columns: SNP ID, gene ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:01<00:00,  1.48it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotypes loaded\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# LOAD HBTRC STUDY DATA\n",
    "from pandas_plink import read_plink1_bin, read_plink\n",
    "import pandas as pd\n",
    "# read_plink1_bin returns (dataArray in shape = (samples, variants))\n",
    "## all information concatenated \n",
    "G =  read_plink1_bin(\"HBTRC data.tsv/Genotypes/AMP-AD_HBTRC_MSSM_IlluminaHumanHap650Y.bed\",\n",
    "                     \"HBTRC data.tsv/Genotypes/AMP-AD_HBTRC_MSSM_IlluminaHumanHap650Y.bim\",\n",
    "                     \"HBTRC data.tsv/Genotypes/AMP-AD_HBTRC_MSSM_IlluminaHumanHap650Y.fam\", \n",
    "                     verbose=False)\n",
    "                     \n",
    "    \n",
    "# read_plink returns (pandasDataFrame, pandasDataFrame, daskArray)\n",
    "# vBim_pNC: allele, vFam_pSC: samples, vBed_aNS: genotype (snp)\n",
    "(vBim_pNC, vFam_pSC, vBed_aNS) = read_plink(\"HBTRC data.tsv/Genotypes/AMP-AD_HBTRC_MSSM_IlluminaHumanHap650Y\")\n",
    "print( \"Genotypes loaded\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a0</th>\n",
       "      <th>a1</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs3094315</td>\n",
       "      <td>0.0</td>\n",
       "      <td>752566</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs12562034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>768448</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs3934834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1005806</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs9442372</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1018704</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs3737728</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1021415</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>rs9442398</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1021695</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>rs6687776</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1030565</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>rs4970405</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1048955</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>rs12726255</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1049950</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>rs11807848</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1061166</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>rs9442373</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1062638</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom         snp   cm      pos a0 a1   i\n",
       "0      1   rs3094315  0.0   752566  C  T   0\n",
       "1      1  rs12562034  0.0   768448  A  G   1\n",
       "2      1   rs3934834  0.0  1005806  T  C   2\n",
       "3      1   rs9442372  0.0  1018704  A  G   3\n",
       "4      1   rs3737728  0.0  1021415  T  C   4\n",
       "5      1   rs9442398  0.0  1021695  A  G   5\n",
       "6      1   rs6687776  0.0  1030565  T  C   6\n",
       "7      1   rs4970405  0.0  1048955  G  A   7\n",
       "8      1  rs12726255  0.0  1049950  G  A   8\n",
       "9      1  rs11807848  0.0  1061166  C  T   9\n",
       "10     1   rs9442373  0.0  1062638  C  A  10"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Columns:\n",
    "\n",
    "Chrom: Chromosome code (either an integer, or 'X'/'Y'/'XY'/'MT'; '0' indicates unknown) or name\n",
    "SNP: Variant identifier\n",
    "Cm: Position in morgans or centimorgans (safe to use dummy value of '0')\n",
    "Pos: Base-pair coordinate (1-based; limited to 2^31-2)\n",
    "A0: Allele 1 (corresponding to clear bits in .bed; usually minor)\n",
    "A1: Allele 2 (corresponding to set bits in .bed; usually major)\n",
    "\"\"\"\n",
    "vBim_pNC.head(11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fid</th>\n",
       "      <th>iid</th>\n",
       "      <th>father</th>\n",
       "      <th>mother</th>\n",
       "      <th>gender</th>\n",
       "      <th>trait</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15876</td>\n",
       "      <td>15876</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15877</td>\n",
       "      <td>15877</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15878</td>\n",
       "      <td>15878</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15879</td>\n",
       "      <td>15879</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15880</td>\n",
       "      <td>15880</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>15881</td>\n",
       "      <td>15881</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>15882</td>\n",
       "      <td>15882</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>15883</td>\n",
       "      <td>15883</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>15884</td>\n",
       "      <td>15884</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>15885</td>\n",
       "      <td>15885</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fid    iid father mother gender  trait  i\n",
       "0  15876  15876      0      0      0   -9.0  0\n",
       "1  15877  15877      0      0      0   -9.0  1\n",
       "2  15878  15878      0      0      0   -9.0  2\n",
       "3  15879  15879      0      0      0   -9.0  3\n",
       "4  15880  15880      0      0      0   -9.0  4\n",
       "5  15881  15881      0      0      0   -9.0  5\n",
       "6  15882  15882      0      0      0   -9.0  6\n",
       "7  15883  15883      0      0      0   -9.0  7\n",
       "8  15884  15884      0      0      0   -9.0  8\n",
       "9  15885  15885      0      0      0   -9.0  9"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vFam_pSC.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fid</th>\n",
       "      <th>iid</th>\n",
       "      <th>father</th>\n",
       "      <th>mother</th>\n",
       "      <th>gender</th>\n",
       "      <th>trait</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16037</th>\n",
       "      <td>16037</td>\n",
       "      <td>16037</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16255</th>\n",
       "      <td>16255</td>\n",
       "      <td>16255</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16240</th>\n",
       "      <td>16240</td>\n",
       "      <td>16240</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16056</th>\n",
       "      <td>16056</td>\n",
       "      <td>16056</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21820</th>\n",
       "      <td>21820</td>\n",
       "      <td>21820</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>739</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         fid    iid father mother gender  trait    i\n",
       "fid                                                 \n",
       "16037  16037  16037      0      0      0   -9.0  735\n",
       "16255  16255  16255      0      0      0   -9.0  736\n",
       "16240  16240  16240      0      0      0   -9.0  737\n",
       "16056  16056  16056      0      0      0   -9.0  738\n",
       "21820  21820  21820      0      0      0   -9.0  739"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make fid as index column\n",
    "vFam_pSC.index = vFam_pSC.loc[ :, \"fid\" ]\n",
    "vFam_pSC.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(740, 672838)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The values of the bed matrix denote how many alleles a1 are in the corresponding position and individual\n",
    "vBed_nSN = vBed_aNS.compute().transpose()\n",
    "vBed_nSN.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[2., 2., 1., ..., 1., 2., 2.],\n",
       "       [2., 2., 2., ..., 1., 2., 2.],\n",
       "       [2., 2., 2., ..., 2., 2., 2.],\n",
       "       ...,\n",
       "       [2., 2., 1., ..., 2., 2., 2.],\n",
       "       [1., 2., 2., ..., 2., 2., 2.],\n",
       "       [2., 2., 2., ..., 2., 2., 2.]])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vBed_nSN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(740,)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dimension checks\n",
    "assert( vBim_pNC.shape[0] == vBed_nSN.shape[1] )\n",
    "assert( vFam_pSC.shape[0] == vBed_nSN.shape[0] )\n",
    "\n",
    "sSubjectsSnps_nN = vFam_pSC['fid'].to_numpy()\n",
    "sSubjectsSnps_nN.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Match snps with corresponding genes\n",
    "### For large datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD ELINK SNP_GENE FILE FROM NCBI \n",
    "import numpy as np\n",
    "sSNP_Gene = np.loadtxt(fname=\"/Elink/snp_genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take snps and genes in separate numpy arrays\n",
    "sSNP_nN = sSNP_Gene[:,0]\n",
    "sGene_nN = sSNP_Gene[:,1]\n",
    "\n",
    "sSNP_nN = sSNP_nN.astype('int64')\n",
    "sGene_nN = sGene_nN.astype('int64')\n",
    "\n",
    "# Sort snp names to make search faster\n",
    "perm = np.argsort(sSNP_nN)\n",
    "sSNP_nN = sSNP_nN[perm]\n",
    "sGene_nN = sGene_nN[perm]\n",
    "\n",
    "# SNP IDs given in the vBim_pNC['snp'] column\n",
    "nNn = np.array( [ ( int(vBim_pNC['snp'][i].replace('rs', ''))) \n",
    "                 for i in range(len(vBim_pNC))]) \n",
    "nNs = np.array( [ vBim_pNC['snp'][i] \n",
    "                 for i in range(len(vBim_pNC))])\n",
    "\n",
    "# FIND IN WHICH GENES GIVEN SNPS ARE\n",
    "nGeneID_0_nG = sGene_nN[np.searchsorted(sSNP_nN, nNn)]\n",
    "sGeneID_0_nG = nGeneID_0_nG.astype('str')\n",
    "\n",
    "assert(sGeneID_0_nG.shape[0] == vBim_pNC.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import RNA expression data\n",
    "## Find if genes are TFs or targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene expressions loaded.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(39579, 473)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import rna expression data\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "sRna_pRC = pd.read_csv(\"HBTRC data.tsv/Gene expression/AMP-AD_HBTRC_MSSM_Agilent44Karray_PFC_AgeCorrected_all.tsv\", \n",
    "                       sep='\\t', \n",
    "                       dtype = np.str,\n",
    "                       na_values = str,\n",
    "                       keep_default_na = False)\n",
    "print(\"Gene expressions loaded.\")\n",
    "sRna_pRC.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reporterid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>clusterName</th>\n",
       "      <th>LocusLinkID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>StartCoordinate</th>\n",
       "      <th>EndCoordinate</th>\n",
       "      <th>expr_Mean</th>\n",
       "      <th>expr_STD</th>\n",
       "      <th>...</th>\n",
       "      <th>16274</th>\n",
       "      <th>16273</th>\n",
       "      <th>16272</th>\n",
       "      <th>16271</th>\n",
       "      <th>16270</th>\n",
       "      <th>16269</th>\n",
       "      <th>16268</th>\n",
       "      <th>16267</th>\n",
       "      <th>16266</th>\n",
       "      <th>16265</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>10020272110</td>\n",
       "      <td>RSE_00000246951</td>\n",
       "      <td>RSE_00000246951</td>\n",
       "      <td>HSG00287442</td>\n",
       "      <td>NA</td>\n",
       "      <td>14</td>\n",
       "      <td>19767400</td>\n",
       "      <td>19767646</td>\n",
       "      <td>0.000909618759585221</td>\n",
       "      <td>0.028563033946755</td>\n",
       "      <td>...</td>\n",
       "      <td>0.019483</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.011418</td>\n",
       "      <td>0.014406</td>\n",
       "      <td>-0.010631</td>\n",
       "      <td>-0.053129</td>\n",
       "      <td>-0.075973</td>\n",
       "      <td>-0.039212</td>\n",
       "      <td>-0.043582</td>\n",
       "      <td>0.009588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>10023804644</td>\n",
       "      <td>Contig48774_RC</td>\n",
       "      <td>Contig48774_RC</td>\n",
       "      <td>HSG00405848</td>\n",
       "      <td>NA</td>\n",
       "      <td>19</td>\n",
       "      <td>47284507</td>\n",
       "      <td>47284972</td>\n",
       "      <td>-0.000811385187482218</td>\n",
       "      <td>0.0753989871525773</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.052873</td>\n",
       "      <td>-0.020108</td>\n",
       "      <td>0.027628</td>\n",
       "      <td>0.022246</td>\n",
       "      <td>-0.030626</td>\n",
       "      <td>-0.057649</td>\n",
       "      <td>0.097587</td>\n",
       "      <td>-0.005713</td>\n",
       "      <td>0.025973</td>\n",
       "      <td>0.035557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>10023804647</td>\n",
       "      <td>NM_002821</td>\n",
       "      <td>PTK7</td>\n",
       "      <td>HSG00243177</td>\n",
       "      <td>5754</td>\n",
       "      <td>6</td>\n",
       "      <td>43152006</td>\n",
       "      <td>43237435</td>\n",
       "      <td>-0.00129990794939302</td>\n",
       "      <td>0.0901277811699355</td>\n",
       "      <td>...</td>\n",
       "      <td>0.105776</td>\n",
       "      <td>-0.062848</td>\n",
       "      <td>0.057068</td>\n",
       "      <td>-0.058405</td>\n",
       "      <td>0.041235</td>\n",
       "      <td>-0.000107</td>\n",
       "      <td>0.121285</td>\n",
       "      <td>0.003693</td>\n",
       "      <td>-0.011672</td>\n",
       "      <td>0.008062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>10023804649</td>\n",
       "      <td>NM_005052</td>\n",
       "      <td>RAC3</td>\n",
       "      <td>HSG00302751</td>\n",
       "      <td>5881</td>\n",
       "      <td>17</td>\n",
       "      <td>77582820</td>\n",
       "      <td>77585368</td>\n",
       "      <td>0.00524912335038691</td>\n",
       "      <td>0.0768838865501222</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.015046</td>\n",
       "      <td>-0.050878</td>\n",
       "      <td>-0.093557</td>\n",
       "      <td>-0.041936</td>\n",
       "      <td>0.073034</td>\n",
       "      <td>0.019985</td>\n",
       "      <td>0.006396</td>\n",
       "      <td>-0.003026</td>\n",
       "      <td>0.00948</td>\n",
       "      <td>0.120577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>10023804650</td>\n",
       "      <td>NM_013302</td>\n",
       "      <td>EEF2K</td>\n",
       "      <td>HSG00294779</td>\n",
       "      <td>29904</td>\n",
       "      <td>16</td>\n",
       "      <td>22125256</td>\n",
       "      <td>22205452</td>\n",
       "      <td>-0.000840262624603848</td>\n",
       "      <td>0.071208396136064</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054709</td>\n",
       "      <td>-0.016822</td>\n",
       "      <td>0.033285</td>\n",
       "      <td>0.002015</td>\n",
       "      <td>-0.045867</td>\n",
       "      <td>-0.080366</td>\n",
       "      <td>0.060596</td>\n",
       "      <td>0.041399</td>\n",
       "      <td>-0.018969</td>\n",
       "      <td>0.023791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>10023804651</td>\n",
       "      <td>NM_001958</td>\n",
       "      <td>EEF1A2</td>\n",
       "      <td>HSG00312077</td>\n",
       "      <td>1917</td>\n",
       "      <td>20</td>\n",
       "      <td>61589809</td>\n",
       "      <td>61600949</td>\n",
       "      <td>-0.00518338465420276</td>\n",
       "      <td>0.0818523427224567</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007156</td>\n",
       "      <td>0.010638</td>\n",
       "      <td>0.043828</td>\n",
       "      <td>0.062611</td>\n",
       "      <td>0.086822</td>\n",
       "      <td>-0.004916</td>\n",
       "      <td>0.003247</td>\n",
       "      <td>-0.004339</td>\n",
       "      <td>0.031675</td>\n",
       "      <td>-0.09015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>10023804653</td>\n",
       "      <td>NM_005309</td>\n",
       "      <td>GPT</td>\n",
       "      <td>HSG00258201</td>\n",
       "      <td>2875</td>\n",
       "      <td>8</td>\n",
       "      <td>145700232</td>\n",
       "      <td>145703357</td>\n",
       "      <td>0.00096687690472571</td>\n",
       "      <td>0.0397133413562591</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.009082</td>\n",
       "      <td>-0.003513</td>\n",
       "      <td>0.032041</td>\n",
       "      <td>-0.001855</td>\n",
       "      <td>0.001469</td>\n",
       "      <td>-0.005628</td>\n",
       "      <td>-0.022021</td>\n",
       "      <td>-0.037238</td>\n",
       "      <td>0.034346</td>\n",
       "      <td>0.005751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>10023804655</td>\n",
       "      <td>NM_019848</td>\n",
       "      <td>SLC10A3</td>\n",
       "      <td>HSG00254520</td>\n",
       "      <td>8273</td>\n",
       "      <td>X</td>\n",
       "      <td>153279351</td>\n",
       "      <td>153282699</td>\n",
       "      <td>0.00111064290061946</td>\n",
       "      <td>0.0387127625853338</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.020499</td>\n",
       "      <td>-0.002079</td>\n",
       "      <td>0.021393</td>\n",
       "      <td>-0.026294</td>\n",
       "      <td>0.006283</td>\n",
       "      <td>-0.02988</td>\n",
       "      <td>0.018761</td>\n",
       "      <td>-0.013479</td>\n",
       "      <td>0.010437</td>\n",
       "      <td>0.018053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>10023804656</td>\n",
       "      <td>NM_001513</td>\n",
       "      <td>GSTZ1</td>\n",
       "      <td>HSG00288811</td>\n",
       "      <td>2954</td>\n",
       "      <td>14</td>\n",
       "      <td>76857458</td>\n",
       "      <td>76867692</td>\n",
       "      <td>-5.82494503961694e-05</td>\n",
       "      <td>0.101248961269102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.058883</td>\n",
       "      <td>0.123517</td>\n",
       "      <td>-0.055339</td>\n",
       "      <td>-0.077121</td>\n",
       "      <td>0.105678</td>\n",
       "      <td>0.093959</td>\n",
       "      <td>0.052151</td>\n",
       "      <td>0.050042</td>\n",
       "      <td>-0.110663</td>\n",
       "      <td>-0.057596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>10023804657</td>\n",
       "      <td>AL110240</td>\n",
       "      <td>RP11-529I10.4</td>\n",
       "      <td>HSG00269139</td>\n",
       "      <td>25911</td>\n",
       "      <td>10</td>\n",
       "      <td>103344414</td>\n",
       "      <td>103359398</td>\n",
       "      <td>-0.000609905034021173</td>\n",
       "      <td>0.0559833434153445</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003699</td>\n",
       "      <td>0.093865</td>\n",
       "      <td>0.040565</td>\n",
       "      <td>0.069427</td>\n",
       "      <td>0.016685</td>\n",
       "      <td>0.030223</td>\n",
       "      <td>-0.034967</td>\n",
       "      <td>0.002586</td>\n",
       "      <td>0.043264</td>\n",
       "      <td>0.024565</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 473 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     reporterid         sequence      gene_symbol  clusterName LocusLinkID  \\\n",
       "40  10020272110  RSE_00000246951  RSE_00000246951  HSG00287442          NA   \n",
       "41  10023804644   Contig48774_RC   Contig48774_RC  HSG00405848          NA   \n",
       "42  10023804647        NM_002821             PTK7  HSG00243177        5754   \n",
       "43  10023804649        NM_005052             RAC3  HSG00302751        5881   \n",
       "44  10023804650        NM_013302            EEF2K  HSG00294779       29904   \n",
       "45  10023804651        NM_001958           EEF1A2  HSG00312077        1917   \n",
       "46  10023804653        NM_005309              GPT  HSG00258201        2875   \n",
       "47  10023804655        NM_019848          SLC10A3  HSG00254520        8273   \n",
       "48  10023804656        NM_001513            GSTZ1  HSG00288811        2954   \n",
       "49  10023804657         AL110240    RP11-529I10.4  HSG00269139       25911   \n",
       "\n",
       "   Chromosome StartCoordinate EndCoordinate              expr_Mean  \\\n",
       "40         14        19767400      19767646   0.000909618759585221   \n",
       "41         19        47284507      47284972  -0.000811385187482218   \n",
       "42          6        43152006      43237435   -0.00129990794939302   \n",
       "43         17        77582820      77585368    0.00524912335038691   \n",
       "44         16        22125256      22205452  -0.000840262624603848   \n",
       "45         20        61589809      61600949   -0.00518338465420276   \n",
       "46          8       145700232     145703357    0.00096687690472571   \n",
       "47          X       153279351     153282699    0.00111064290061946   \n",
       "48         14        76857458      76867692  -5.82494503961694e-05   \n",
       "49         10       103344414     103359398  -0.000609905034021173   \n",
       "\n",
       "              expr_STD  ...      16274      16273      16272      16271  \\\n",
       "40   0.028563033946755  ...   0.019483   0.000138   0.011418   0.014406   \n",
       "41  0.0753989871525773  ...  -0.052873  -0.020108   0.027628   0.022246   \n",
       "42  0.0901277811699355  ...   0.105776  -0.062848   0.057068  -0.058405   \n",
       "43  0.0768838865501222  ...  -0.015046  -0.050878  -0.093557  -0.041936   \n",
       "44   0.071208396136064  ...   0.054709  -0.016822   0.033285   0.002015   \n",
       "45  0.0818523427224567  ...  -0.007156   0.010638   0.043828   0.062611   \n",
       "46  0.0397133413562591  ...  -0.009082  -0.003513   0.032041  -0.001855   \n",
       "47  0.0387127625853338  ...  -0.020499  -0.002079   0.021393  -0.026294   \n",
       "48   0.101248961269102  ...   0.058883   0.123517  -0.055339  -0.077121   \n",
       "49  0.0559833434153445  ...  -0.003699   0.093865   0.040565   0.069427   \n",
       "\n",
       "        16270      16269      16268      16267      16266      16265  \n",
       "40  -0.010631  -0.053129  -0.075973  -0.039212  -0.043582   0.009588  \n",
       "41  -0.030626  -0.057649   0.097587  -0.005713   0.025973   0.035557  \n",
       "42   0.041235  -0.000107   0.121285   0.003693  -0.011672   0.008062  \n",
       "43   0.073034   0.019985   0.006396  -0.003026    0.00948   0.120577  \n",
       "44  -0.045867  -0.080366   0.060596   0.041399  -0.018969   0.023791  \n",
       "45   0.086822  -0.004916   0.003247  -0.004339   0.031675   -0.09015  \n",
       "46   0.001469  -0.005628  -0.022021  -0.037238   0.034346   0.005751  \n",
       "47   0.006283   -0.02988   0.018761  -0.013479   0.010437   0.018053  \n",
       "48   0.105678   0.093959   0.052151   0.050042  -0.110663  -0.057596  \n",
       "49   0.016685   0.030223  -0.034967   0.002586   0.043264   0.024565  \n",
       "\n",
       "[10 rows x 473 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sGeneSymbols_nR = sRna_pRC['gene_symbol'].to_numpy()\n",
    "sRna_pRC[40:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load transcription factors and target genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Set up environment variables for rpy2 to work\n",
    "os.environ['R_HOME'] = '/usr/local/Cellar/r/3.6.0_2/lib/R'\n",
    "os.environ['R_USER'] = '/usr/local/lib/python3.7/site-packages/rpy2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data loaded from Rdata file.\n",
      "Variables loaded:\n",
      "targets_iTiC\n",
      "tfs_iFiC\n"
     ]
    }
   ],
   "source": [
    "# Load Rdata with rpy2\n",
    "import rpy2\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "pandas2ri.activate()\n",
    "# Clear the workspace\n",
    "r( 'rm( list = ls( ) )' )\n",
    "# 16GB memory limit\n",
    "r( 'memory.limit( 16000 )' )\n",
    "r['load']('gwasPreprocessing_v2_tfs.Rdata')\n",
    "print( 'Data loaded from Rdata file.' )\n",
    "\n",
    "# Hale the user\n",
    "print( 'Variables loaded:' )\n",
    "[ print( a ) for a in r['ls']() ];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "##tfs_iFiC\n",
    "sTFs_nPC = np.array( rpy2.robjects.r( 't( tfs_iFiC )' ),\n",
    "                    dtype = np.str )\n",
    "sTFs_nPC.shape = r( 'dim( tfs_iFiC )' )\n",
    "\n",
    "## targets_iTiC\n",
    "sTargets_nPC = np.array( rpy2.robjects.r( 't( targets_iTiC )' ),\n",
    "                    dtype = np.str )\n",
    "sTargets_nPC.shape = r( 'dim( targets_iTiC )' )\n",
    "\n",
    "## targets_iTiC[, \"tf_symbol\"]\n",
    "sTFs_nP = np.array( r( ' targets_iTiC[, \"tf_symbol\"] ' ) )\n",
    "\n",
    "## targets_iTiC[, \"target_symbol\"]\n",
    "sTargets_nP = np.array( r( ' targets_iTiC[, \"target_symbol\"] ' ) )\n",
    "\n",
    "## Find TFs and targets given in sRna_pRS['gene_symbol']\n",
    "nTFs_nR = np.array([sGeneSymbols_nR[i] in sTFs_nP for i in range(sGeneSymbols_nR.shape[0])])\n",
    "sTFs_nT = sGeneSymbols_nR[np.where(nTFs_nR == True)[0]]\n",
    "\n",
    "nTargets_nR = np.array([sGeneSymbols_nR[i] in sTargets_nP for i in range(sGeneSymbols_nR.shape[0])])\n",
    "sTargets_nT = sGeneSymbols_nR[np.where(nTargets_nR == True)[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "611\n"
     ]
    }
   ],
   "source": [
    "%run loadData_utils.py\n",
    "## Create pairs of transcription factors and targets\n",
    "nPairs_lP = []\n",
    "for tf in sTFs_nT:\n",
    "    nTFtargetList = tf_Pairs_Finding(tf, \n",
    "                                     sTFs_nP, \n",
    "                                     sTargets_nP, \n",
    "                                     sGeneSymbols_nR)\n",
    "    nPairs_lP.extend(nTFtargetList)\n",
    "\n",
    "nPairs_nP = np.array(nPairs_lP)\n",
    "nPairsTT_nP = sTargets_nPC[nPairs_nP]\n",
    "nPairs_lPlC = [nPairsTT_nP[:,i] for i in range(5) ]\n",
    "\n",
    "## Find unique genes that are TFs or targets in order to filter RNA expression data\n",
    "sGenes_lG = []\n",
    "sTFs_lT = np.unique(nPairs_lPlC[1]).tolist()\n",
    "sTargets_lT = np.unique(nPairs_lPlC[3]).tolist()\n",
    "sGenes_lG.extend(sTFs_lT)\n",
    "sGenes_lG.extend(sTargets_lT)\n",
    "uGenes_lG = np.unique(sGenes_lG).tolist()\n",
    "print(len(uGenes_lG))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overlapped number of subjects are:  434\n",
      "Process done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.7/site-packages/ipykernel_launcher.py:36: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike\n"
     ]
    }
   ],
   "source": [
    "# Check which subjects overlap in rna expression data and snps data\n",
    "cols_excluded = [\"reporterid\", \"sequence\", \"gene_symbol\", \"clusterName\", \"LocusLinkID\", \"Chromosome\", \"StartCoordinate\", \"EndCoordinate\", \"expr_Mean\", \"expr_STD\"]\n",
    "sRna_0_pRS = sRna_pRC.drop(columns = cols_excluded, axis = 1)\n",
    "sSubjectsRna_nR = sRna_0_pRS.columns.to_numpy()\n",
    "\n",
    "nSubjectRnaIndexinSnps_nR = np.array( [np.where(sSubjectsRna_nR[i] == sSubjectsSnps_nN )[0][0]\n",
    "                                       if len(np.where(sSubjectsRna_nR[i] == sSubjectsSnps_nN )[0]) != 0 else -1\n",
    "                                       for i in range(sSubjectsRna_nR.shape[0]) ] ) \n",
    "\n",
    "sSubjectsRna_nR = np.delete(sSubjectsRna_nR, np.where(nSubjectRnaIndexinSnps_nR == -1)[0])\n",
    "\n",
    "nSubjectRnaIndexinSnps_lR = np.delete(nSubjectRnaIndexinSnps_nR, np.where(nSubjectRnaIndexinSnps_nR == -1)[0]).tolist()\n",
    "\n",
    "print(\"Overlapped number of subjects are: \",  sSubjectsRna_nR.shape[0])\n",
    "\n",
    "\n",
    "# Select only overlapped subjects for vBed_nSN (genotypes), vFam_pSC (samples) and sRna_pRS (rna expression)\n",
    "rSnp_t_nSN = vBed_nSN[nSubjectRnaIndexinSnps_lR]\n",
    "rSnp_nSN = rSnp_t_nSN\n",
    "rSnp_nSN[np.isnan(rSnp_t_nSN)] = 3\n",
    "rSnp_nSN[rSnp_t_nSN == 2] = 3\n",
    "rSnp_nSN[rSnp_t_nSN == 1] = 2\n",
    "rSnp_nSN[rSnp_t_nSN == 0] = 1\n",
    "genotype_categories = np.unique(rSnp_nSN)\n",
    "assert(len(genotype_categories) == 3)\n",
    "\n",
    "vFam_pSC = vFam_pSC.loc[sSubjectsRna_nR.tolist()]\n",
    "sRna_1_pRS = sRna_0_pRS[sSubjectsRna_nR.tolist()]\n",
    "nSs = sSubjectsRna_nR # subject names\n",
    "\n",
    "assert( vFam_pSC.shape[0] == rSnp_nSN.shape[0] )\n",
    "assert( vFam_pSC.shape[0] == sRna_1_pRS.shape[1] )\n",
    "assert( rSnp_nSN.shape[0] == sRna_1_pRS.shape[1] )\n",
    "\n",
    "sRna_2_pRS = sRna_1_pRS.set_index(sRna_pRC['LocusLinkID'])\n",
    "sRna_3_pRS = sRna_2_pRS.loc[uGenes_lG]\n",
    "print(\"Process done.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create rRna_nSR, nPairs_lPlC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rRna_nSR is created.\n",
      "Genes with missing values are dropped.\n",
      "NA values are replaced with average of rows.\n",
      "nPairs_lPlC is updated.\n"
     ]
    }
   ],
   "source": [
    "# Create transcripts\n",
    "nRs = np.array( uGenes_lG )\n",
    "\n",
    "# Find duplicate RNA expressions of each gene\n",
    "sRna_lRlC = sRna_3_pRS.index.tolist()\n",
    "duplicateGenes = 1\n",
    "\n",
    "nIndex_lR = [duplicateGenes]\n",
    "\n",
    "for iG in range(len(sRna_lRlC)-1):\n",
    "    if sRna_lRlC[iG + 1 ] != sRna_lRlC[iG]:\n",
    "        duplicateGenes += 1\n",
    "    nIndex_lR.append(duplicateGenes)\n",
    "    \n",
    "nIndex_nR = np.array(nIndex_lR)\n",
    "\n",
    "# Create a new pandas dataframe for average of RNA expression if there are multiple values of each gene\n",
    "rRna_pRS = pd.DataFrame(columns=sSubjectsRna_nR)\n",
    "\n",
    "for iR in range(len(nRs)):\n",
    "    index = np.where(nIndex_nR == (iR + 1))[0]\n",
    "    gene = nRs[iR]    \n",
    "    sRna_4_pRS = sRna_3_pRS.iloc[index]\n",
    "    sRna_4_pRS = sRna_4_pRS.replace(\"NA\", np.nan)\n",
    "    \n",
    "    sRna_4_pRS = sRna_4_pRS.astype('float')\n",
    "    rRna_pRS.loc[gene] = sRna_4_pRS.mean(axis = 0)\n",
    "\n",
    "\n",
    "rRna_pRS = rRna_pRS.dropna(how='all')\n",
    "rRna_nSR = rRna_pRS.transpose().to_numpy()\n",
    "assert( rRna_nSR.shape[0] == rSnp_nSN.shape[0])\n",
    "print(\"rRna_nSR is created.\")\n",
    "\n",
    "# Find and drop NA rows\n",
    "nNA_nIndex = np.unique(np.where(sRna_3_pRS.isna())[0])\n",
    "sNA_nGenes = sRna_3_pRS.iloc[nNA_nIndex].index.to_numpy()\n",
    "nNA_lIndex = [np.where(sNA_nGenes[i] == nRs)[0][0] for i in range(len(sNA_nGenes))]\n",
    "nRs = np.delete(nRs, nNA_lIndex)\n",
    "\n",
    "# Create transcripts' corresponding SNP pairs\n",
    "nNs_lRlN = [nNs[np.where(iR == sGeneID_0_nG)] for iR in nRs]\n",
    "assert( len(nRs) == len(nNs_lRlN) )\n",
    "\n",
    "print(\"Genes with missing values are dropped.\")\n",
    "assert( rRna_nSR.shape[1] == len(nRs) )\n",
    "\n",
    "# Replace NA values with average of rows\n",
    "row_means = np.nanmean(rRna_nSR, axis = 1) \n",
    "na_indices = np.where(np.isnan(rRna_nSR))\n",
    "rRna_nSR[na_indices] = np.take(row_means, na_indices[0])\n",
    "assert (len(np.where(np.isnan(rRna_nSR))[0]) == 0)\n",
    "assert (len(np.where(np.isnan(rRna_nSR))[1]) == 0)\n",
    "print(\"NA values are replaced with average of rows.\")\n",
    "\n",
    "# Update nPairs_lPlC with removed genes\n",
    "nNA_Index_Pairs = np.array( [np.where(sNA_nGenes[i] == nPairs_lPlC[3])[0][0] for i in range(len(sNA_nGenes))] )\n",
    "nPairs_lPlC = [np.delete(nPairs_lPlC[iC], nNA_Index_Pairs) for iC in range(len(nPairs_lPlC)) ] \n",
    "print(\"nPairs_lPlC is updated.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eQTL inputs in pandas format created.\n",
      "eQTL inputs stored in CSV format.\n",
      "eQTL input helpers stored in CSV format.\n"
     ]
    }
   ],
   "source": [
    "# CREATE INPUTS FOR EQTL ANALYSIS\n",
    "link_genotype = np.zeros(shape = rSnp_nSN.shape)\n",
    "link_genotype[rSnp_nSN == 1] = 0\n",
    "link_genotype[rSnp_nSN == 2] = 1\n",
    "link_genotype[rSnp_nSN == 3] = 2\n",
    "\n",
    "# GENOTYPE\n",
    "eqtl_genotype = pd.DataFrame(data = rSnp_nSN.transpose(), columns=nSs)\n",
    "eqtl_genotype.set_index(nNs, inplace=True)\n",
    "eqtl_genotype.index.rename(\"snp\", inplace=True)\n",
    "\n",
    "# GENE EXPRESSION\n",
    "eqtl_expression = pd.DataFrame(data = rRna_nSR.transpose(), columns = nSs)\n",
    "eqtl_expression.set_index(nRs, inplace=True)\n",
    "eqtl_expression.index.rename(\"id\", inplace=True)\n",
    "\n",
    "# LINKAGE DISEQUILIBRIUM GENOTYPE\n",
    "linkdis_genotype = pd.DataFrame(data = link_genotype.transpose(), columns=nSs)\n",
    "linkdis_genotype.set_index(nNs, inplace=True)\n",
    "linkdis_genotype.index.rename(\"snp\", inplace=True)\n",
    "\n",
    "# PROBE_LOC: GENE LOCATIONS\n",
    "eqtl_probe_lob_cols = ['LocusLinkID', 'Chromosome', 'StartCoordinate', 'EndCoordinate']\n",
    "eqtl_probe_lob_0 = sRna_pRC[probe_lob_cols]\n",
    "eqtl_probe_lob_1 = eqtl_probe_lob_0.set_index(probe_lob_0['LocusLinkID'])\n",
    "eqtl_probe_lob_1 = eqtl_probe_lob_1.loc[nRs]\n",
    "eqtl_probe_lob_1.drop_duplicates(subset='LocusLinkID', keep='first', inplace=True)\n",
    "eqtl_probe_lob_1.drop(columns=['LocusLinkID'], inplace=True)\n",
    "eqtl_probe_lob_1.index.rename('id', inplace=True)\n",
    "\n",
    "# SNP_LOC: SNP POSITIONS\n",
    "snp_loc_data = pd.DataFrame{'snp': vBim_pNC['snp'], \n",
    "                            'chrom_snp': vBim_pNC['chrom'],\n",
    "                            'pos': vBim_pNC['pos'] }\n",
    "\n",
    "eqtl_snp_loc = pd.DataFrame(data = snp_loc_data)\n",
    "eqtl_snp_loc.set_index(nNs, inplace=True)\n",
    "eqtl_snp_loc.drop(columns=['snp'], inplace=True)\n",
    "eqtl_snp_loc.index.rename(\"snp\", inplace=True)\n",
    "\n",
    "# MAPPING from SNP to GENE\n",
    "snp_gene_data = {'snp': nNs, 'gene': sGeneID_0_nG}\n",
    "snp_gene = pd.DataFrame(data = snp_gene_data)\n",
    "\n",
    "assert( eqtl_genotype.shape[1] == eqtl_expression.shape[1] )\n",
    "assert( eqtl_genotype.shape[0] == eqtl_snp_loc.shape[0] )\n",
    "assert( eqtl_expression.shape[0] == eqtl_probe_lob_1.shape[0] )\n",
    "assert( eqtl_expression.index.name == eqtl_probe_lob_1.index.name)\n",
    "assert( eqtl_genotype.index.name == eqtl_snp_loc.index.name)\n",
    "\n",
    "print(\"eQTL inputs in pandas format created.\")\n",
    "\n",
    "# eQTL: STORE IN .CSV FORMAT\n",
    "expression_csv = eqtl_expression.to_csv(\"../eQTL + LD analysis/HBTRC/eqtl_expression.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "genotype_csv = eqtl_genotype.to_csv(\"../eQTL + LD analysis/HBTRC/eqtl_genotype.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "probe_loc_csv = eqtl_probe_lob_1.to_csv(\"../eQTL + LD analysis/HBTRC/eqtl_probe_loc.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "snp_loc_csv = eqtl_snp_loc.to_csv(\"../eQTL + LD analysis/HBTRC/eqtl_snp_loc.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "print(\"eQTL inputs stored in CSV format.\")\n",
    "\n",
    "# LINKAGE DISEQUILIBRIUM: STORE IN .CSV FORMAT\n",
    "linkdis_genotype_csv = linkdis_genotype.to_csv(\"../eQTL + LD analysis/HBTRC/linkdis_genotype.csv\",\n",
    "                                        sep=\",\",\n",
    "                                        header=True)\n",
    "\n",
    "# HELPER VARIABLES\n",
    "subjects_csv = pd.DataFrame(data=nSs).to_csv(\"../eQTL + LD analysis/HBTRC/eqtl_subjects.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "snps_csv = pd.DataFrame(data=nNs).to_csv(\"../eQTL + LD analysis/HBTRC/eqtl_snps.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "genes_csv = pd.DataFrame(data=nRs).to_csv(\"../eQTL + LD analysis/HBTRC/eqtl_genes.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "snp_gene_csv = pd.DataFrame(data=snp_gene).to_csv(\"../eQTL + LD analysis/HBTRC/eqtl_snp_gene.csv\",\n",
    "                                   sep =\",\",\n",
    "                                   header=True)\n",
    "\n",
    "print(\"eQTL input helpers stored in CSV format.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data saved into pickle.\n"
     ]
    }
   ],
   "source": [
    "# SAVE DATA\n",
    "# Save all the transformed data into a pickle, such that it can be more easily loaded from Python\n",
    "\n",
    "# Save data into Python file\n",
    "import pickle\n",
    "with open( 'loadData_HBTRC.pickle', 'wb' ) as f:\n",
    "    pickle.dump( rRna_nSR, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( rSnp_nSN, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nNs_lRlN, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nPairs_lPlC, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nRs, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nNs, f, pickle.HIGHEST_PROTOCOL )\n",
    "    pickle.dump( nSs, f, pickle.HIGHEST_PROTOCOL )\n",
    "    print( 'Data saved into pickle.' )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
